### ISSUE TITLE
How are features shuffled around after the processors? Bug with applying sorted features twice?

### ISSUE THREAD

**Issue by davidsc-unity3d** [facebookresearch/ReAgent#211]
Building off the DQN workflow for the cartpole example, I am trying to track what the normalization is doing to my features. One learning is that the integer names for the features need to 0-up, not 1-up, or else [PythonSparseToDenseProcessor.process](https://github.com/facebookresearch/ReAgent/blob/master/ml/rl/preprocessing/sparse_to_dense.py#L67) 
is going to add a column of missing values.

But the confusing thing is, it seems that the sorting feature list is applied twice. 
The first time is in [PandasSparseToDenseProcessor.process](https://github.com/facebookresearch/ReAgent/blob/master/ml/rl/preprocessing/sparse_to_dense.py#L43)

where the feature_to_index is used

        for sd in sparse_data:
            sd_int = {}
            for k, v in sd.items():
                sd_int[self._sparse_to_dense.feature_to_index[int(k)]] = v
 
and the second time is in when [PythonSparseToDenseProcessor(.process](https://github.com/facebookresearch/ReAgent/blob/master/ml/rl/preprocessing/sparse_to_dense.py#L68)
constructs the torch vector:

        values = torch.from_numpy(
            state_features_df[self.sorted_features].values
        ).float()

using sorted_features - this seems like a bug?

I want to plot a before and after of the what the normalization does - but it is very confusing how to line things up. That is, I'd like to write some code like

```
batch_preprocessor = DiscreteDqnBatchPreprocessor(Preprocessor(state_normalization, False))

sorted_features, _ = sort_features_by_normalization(state_normalization)

preprocess_handler = DiscreteDqnPreprocessHandler(
    len(action_names), PandasSparseToDenseProcessor(sorted_features)
)

train_dataset = JSONDatasetReader(
    params["training_data_path"],
    batch_size=training_parameters.minibatch_size,
    preprocess_handler=preprocess_handler,
)

train_dataset.reset_iterator()
data_streamer = DataStreamer(train_dataset, pin_memory=False)

raw_batches = []
batches = []
for idx, batch in enumerate(data_streamer):
    raw_batches.append(batch)
    preprocessed_batch = batch_preprocessor(batch)
    batches.append(preprocessed_batch)

X = np.concatenate([pre_b.training_input.state.float_features.numpy() for pre_b in batches])
raw_X = np.concatenate([b.training_input.state.float_features.value.numpy() for b in raw_batches])
```
and then compare columns - but identifying where a column in raw_X when to X is tricky.

**Comment by czxttkl** [comment]
Let's first discuss the bug. 
So you think that sd_int is keyed by feature index while in PythonSparseToDenseProcessor we still sort state_features_df by feature id?

**Comment by davidsc-unity3d** [comment]
What I'm finding is that my feature '1' initially gets sorted to position 8, that is the output of the `sort_features_by_normalization` -- that is the feature index as you say. A pandas dataframe is made from that, this already has integer column names. But the final PythonSparseToDense effectively applies the feat2idx again - so now feature '1' has been moved to position 15 (in my case). 

If I'm trying to find where that feature went - I would assume I would just look at the sorted_features list once, and assume it is in position 8. That's why it seems like a bug - that list doesn't tell me where it finally went. 

What I haven't checked yet is if the correct normalization is applied - a bigger bug would be if the normalization for '1' is applied to column 8 and not 15

**Comment by czxttkl** [comment]
You can use batch_preprocessor.state_preprocessor.sorted_feature_boundaries to get each feature's start and end index. This way you can verify whether we have a bug. I suspect we do.

**Comment by davidsc-unity3d** [comment]
Thanks, from reading the code, it looks like the state_preprocessor applies the normalizations to the wrong columns - I changed that line in PandasSparseToDenseProcessor
```

--- a/ml/rl/preprocessing/sparse_to_dense.py
+++ b/ml/rl/preprocessing/sparse_to_dense.py
@@ -66,7 +66,8 @@ class PythonSparseToDenseProcessor(SparseToDenseProcessor):
             if col not in state_features_df.columns:
                 state_features_df[col] = missing_value
         values = torch.from_numpy(
-            state_features_df[self.sorted_features].values
+             state_features_df.values
         ).float()
         if self.set_missing_value_to_zero:

```
and I'm much happier, plotting before/after of each column, the plots make sense now. 

I don't know if it is safe to suggest you make that change - if there are two things involved in the chain to reading the batch, in this case

* PandasSparseToDense
* PythonSparseToDense

Then only one of them should be shuffling the parameters, I'm not sure which one - for example you said you lack a test for PandasSparseToDense - maybe the intention was that should not have shuffled? And the shuffling should occur in the PythonSparseToDense? Maybe the idea is Pandas or other - that just gets the data ready for the PythonSparseToDense - and that is always called and should own the shuffling? In anycase, I don't know the code well enough here.

**Comment by czxttkl** [comment]
Fixed
